squint.c

```diff
@@ -0,0 +1,1611 @@
+               *scanp1 = (*scanp1 & 0x0fffffff) | (*scanm1 & 0xf0000000);
+            }
+            else {
+               *scanp1 = (*scanp1 & 0x0fffffff) | (*scanm2 & 0xf0000000);
+            }
+            *scanm2 = NOP;
+            *scanm1 = NOP;
+            *scan   = NOP;
+         }
+      }
+   }
+}
+
+/* "bcond target1 ; b target2 ; taget1: <inst>;" -> "bnotcond target2;" */
+/* todo: may want to check for chains of intervening nops here */
+void simplify_branch5(int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, S_FWD);
+      if ((*scan & 0x0fffffff) == 0x0a000000 &&  // cond jump past next inst
+          ((*scan >> 28) &0xf) < 0xe &&
+          (scan[1] & 0xff000000) == 0xea000000) {
+         scan[1] = (scan[0] & 0xff000000) | (scan[1] & 0x00ffffff);
+         scan[1] = scan[1] ^ (int) 0x10000000;
+         scan[0] = NOP;
+      }
+   }
+}
+
+
+/* optimize branch-specific code sequences */
+void simplify_branch(int *funcBegin, int *funcEnd) {
+   simplify_branch1(funcBegin, funcEnd);
+   simplify_branch2(funcBegin, funcEnd);
+   simplify_branch3(funcBegin, funcEnd);
+   simplify_branch4(funcBegin, funcEnd);
+   simplify_branch5(funcBegin, funcEnd);
+   simplify_branch1(funcBegin, funcEnd); // optional
+}
+
+/* remove all NOP instructions ( mov r0, r0 ) and adjust branches */
+/* mode == 0: intra-function repack, mode != 0, inter-function */
+int *relocate_nop(int *funcBegin, int *funcEnd, int mode)
+{
+   int *retVal = funcEnd;
+   int *memblk;
+   int *branchAddr;
+   int *branchTarget;
+   int *permutation;
+   int ii, jj, tmp, done;
+   int currAddr;
+   int currTarget;
+   int hasLink;
+   int offset;
+   int align;
+
+   int *scan;
+   int *packed;
+
+   /* count number of branches to relocate */
+   int nopCount = 0;
+   int branchCount = 0;
+
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      while (is_const(scan)) ++scan;
+
+      // legal since rename_nop occurs before this call...
+      if (*scan == NOP) {
+         ++nopCount;
+      }
+      else if ((*scan & 0x0e000000) == 0x0a000000) {
+         hasLink = *scan & (1<<24);
+         if ( ((mode == 0) && !hasLink) ||   // intra-function mode
+              ((mode != 0) &&  hasLink) ) {  // inter-function mode
+            ++branchCount;
+         }
+      }
+   }
+
+   if (branchCount != 0) {
+      memblk = malloc(3*branchCount*4);
+      branchAddr = memblk;
+      branchTarget = branchAddr + branchCount;
+      permutation = branchTarget + branchCount;
+
+      /* record all branch inst and target addresses */
+      branchCount = 0;
+      for (scan = funcBegin; scan < funcEnd; ++scan) {
+         while (is_const(scan)) ++scan;
+
+         if ((*scan & 0x0e000000) == 0x0a000000) {
+            hasLink = *scan & (1<<24);
+            if ( ((mode == 0) && !hasLink) ||   // intra-function mode
+                 ((mode != 0) &&  hasLink) ) {  // inter-function mode
```

Comment:
I am not sure what do you mean with the term `intra-function` and `inter-function`.

I guess that by `intra-function` you mean that NOP removal will only be applied to basic blocks within a function and the function address won't be changed; OTOH, by `inter-function` you mean that NOP removal will be applied across functions (in addition to the basic block).

If my "guess" is correct, shouldn't the condition:

```
            if ( ((mode == 0) && !hasLink) ||   // intra-function mode
                 ((mode != 0) &&  hasLink) ) {  // inter-function mode
```

be:


```
            if ( !hasLink ||   // basic block branch for intra/inter-function mode
                 ((mode != 0) &&  hasLink) ) {  // inter-function mode
```

Reply:
If you use scripts/disasm to compare the code before applying Squint and after Squint, you will see large NOP blocks between functions.  The inter-procedual call to relocate_nop removes those NOP blocks between functions (after I fix the bug, which is on my list, but at low priority right now).  The intra-function call to relocate-NOP takes the NOPS produced by peephole optimization out of a given function.